jax/lax_linalg.py

```diff
@@ -319,18 +320,28 @@ def triangular_solve_jvp_rule_a(
     g_a, ans, a, b, left_side, lower, transpose_a, conjugate_a, unit_diagonal):
   if g_a is ad_util.zero:
     return ad_util.zero
+  m, n = b.shape[-2:]
   k = 1 if unit_diagonal else 0
   g_a = np.tril(g_a, k=-k) if lower else np.triu(g_a, k=k)
   g_a = lax.neg(g_a)
   g_a = np.swapaxes(g_a, -1, -2) if transpose_a else g_a
   g_a = np.conj(g_a) if conjugate_a else g_a
-  tmp = triangular_solve(a, g_a, left_side, lower, transpose_a, conjugate_a,
-                         unit_diagonal)
   dot = lax.dot if g_a.ndim == 2 else lax.batch_matmul
+
+  def a_inverse(rhs):
+    return triangular_solve(a, rhs, left_side, lower, transpose_a, conjugate_a,
+                            unit_diagonal)
+
   if left_side:
-    return dot(tmp, ans)
+    # Note: triangular_solve is about the same cost as matrix multplication
+    # (~n^2 FLOPs for matrix/vector inputs). Order these operations in
+    # whichever order is cheaper.
+    if m > n:
+      return a_inverse(dot(g_a, ans))  # A^{-1} (∂A X)
+    else:
+      return dot(a_inverse(g_a), ans)  # (A^{-1} ∂A) X
   else:
-    return dot(ans, tmp)
+    return dot(ans, a_inverse(g_a))
```

Comment:
Doesn't the same apply in the `left_side == False` case?

Reply:
Yes, I suspect so, though I was not confident how the math works out to compute `dot(ans, a_inverse)` in a matrix-free way. Given that `left_side=True` is most common, maybe I'll just add a TODO note here for now?

Reply:
Actually, this turned out to be pretty straightforward once I thought about it. Please take another look.

In particular, I'm a little nervous about test coverage. We don't have any direct test coverage for triangular_solve's JVP (the solve_triangular tests only cover `left_side=True`). There's some indirectly coverage via the JVP rules for cholesky and LU -- maybe that's good enough?

Reply:
Well, more testing is better.

I think the only reason the test isn't there is that `scipy`'s API doesn't expose this option. (We expose it mostly because LAPACK does, and it's sometimes pretty handy.)

We could test the primitive directly instead of or in addition to the `scipy` wrapper.

Up to you.

Reply:
I switched over the gradient tests to use `lax_linalg`